processors/ratelimiter/factory.go

```diff
@@ -0,0 +1,90 @@
+package ratelimiter
+
+import (
+	"context"
+	"net"
+	"strconv"
+	"time"
+
+	pb "github.com/envoyproxy/go-control-plane/envoy/service/ratelimit/v3"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+const (
+	typeStr                         = "hypertrace_ratelimiter"
+	defaultServiceHost              = "127.0.0.1"
+	defaultServicePort              = uint16(8081)
+	defaultDomain                   = "collector"
+	defaultDomainSoftLimitThreshold = uint32(100000) // Soft limit kicks in when limit remaining under 100k
+	defaultHeaderName               = "x-tenant-id"
+	defaultTimeoutMillis            = uint32(1000) // 1 second
+)
+
+// NewFactory creates a factory for the ratelimit processor.
+func NewFactory() component.ProcessorFactory {
+	return component.NewProcessorFactory(
+		typeStr,
+		createDefaultConfig,
+		component.WithTracesProcessor(createTraceProcessor, component.StabilityLevelStable),
+	)
+}
+
+func createDefaultConfig() config.Processor {
+	return &Config{
+		ProcessorSettings: config.NewProcessorSettings(
+			config.NewComponentID(typeStr),
+		),
+		ServiceHost:                  defaultServiceHost,
+		ServicePort:                  defaultServicePort,
+		Domain:                       defaultDomain,
+		DomainSoftRateLimitThreshold: defaultDomainSoftLimitThreshold,
+		TenantIDHeaderName:           defaultHeaderName,
+		TimeoutMillis:                defaultTimeoutMillis,
+	}
+}
+
+func createTraceProcessor(
+	ctx context.Context,
+	params component.ProcessorCreateSettings,
+	cfg config.Processor,
+	nextConsumer consumer.Traces,
+) (component.TracesProcessor, error) {
+	pCfg := cfg.(*Config)
+	rateLimitServiceClient, rateLimitServiceClientConn, cancelFunc, err := getRateLimitServiceClient(ctx, pCfg.ServiceHost, pCfg.ServicePort, pCfg.TimeoutMillis, params)
```

Comment:
If you have declared default config value, are they merged into the incoming config? I don't think so but I could be wrong.

Reply:
You are right, not able to find where config merge happening, can you share reference for this ?

Reply:
I think you have to check for empty value manually and set them yourself.

Reply:
Not sure how it is happening but config override is happening  if it is in config.yaml. I have tested locally.

Reply:
Not config override. If you do not have timeoutmillis set, what's the config value passed on to the processor? Is it the default value of 1000ms?

Reply:
Yes.